flips/component-interface.md

```diff
@@ -0,0 +1,364 @@
+```golang
+// Startable provides an interface to start a component. Once started, the component
+// can be stopped by cancelling the given context.
+type Startable interface {
+  // Start starts the component. Any errors encountered during startup should be returned
+  // directly, whereas irrecoverable errors encountered while the component is running
+  // should be thrown with the given SignalerContext.
+  // This method should only be called once, and subsequent calls should return ErrMultipleStartup.
+  Start(irrecoverable.SignalerContext) error
+}
+```
+Components which implement this interface are passed in a `SignalerContext` upon startup, which they can use to propagate any irrecoverable errors they encounter up to their parent via `SignalerContext.Throw`. The parent can then choose to handle these errors however they like, including restarting the component, logging the error, propagating the error to their own parent, etc.
+
+```golang
+// We define a constrained interface to provide a drop-in replacement for context.Context
+// including in interfaces that compose it.
+type SignalerContext interface {
+  context.Context
+  Throw(err error) // delegates to the signaler
+}
+
+// Signaler sends the error out.
+type Signaler struct {
+  errChan   chan error
+  errThrown *atomic.Bool
+}
+
+func NewSignaler() *Signaler {
+  return &Signaler{
+    errChan:   make(chan error, 1),
+    errThrown: atomic.NewBool(false),
+  }
+}
+
+// Error returns the Signaler's error channel.
+func (s *Signaler) Error() <-chan error {
+  return s.errChan
+}
+
+// Throw is a narrow drop-in replacement for panic, log.Fatal, log.Panic, etc
+// anywhere there's something connected to the error channel. It only sends
+// the first error it is called with to the error channel, and there are various
+// options as to how subsequent errors can be handled.
+func (s *Signaler) Throw(err error) {
+  defer runtime.Goexit()
+
+  // We only propagate the first irrecoverable error to the parent
+  if s.errThrown.CAS(false, true) {
+    s.errChan <- err
+    close(s.errChan)
+  } else {
+    // Another thread, possibly from the same component, has already thrown
+    // an irrecoverable error to this Signaler. Any subsequent irrecoverable
+    // errors can either be logged or ignored, as the parent will already
+    // be taking steps to remediate the first error.
+  }
+}
+```
+
+> For more details about `SignalerContext` and `ErrMultipleStartup`, see [#1275](https://github.com/onflow/flow-go/pull/1275) and [#1355](https://github.com/onflow/flow-go/pull/1355/).
+
+To start a component, a `SignalerContext` must be created to start it with:
+
+```golang
+// this is the context for the routine which manages the component
+var parentCtx context.Context
+
+ctx, cancel := context.WithCancel(parentCtx)
+signaler := irrecoverable.NewSignaler()
+signalerCtx := irrecoverable.WithSignaler(ctx, signaler)
+
+go func() {
+  select {
+  case err := <-signaler.Error():
```

Comment:
We have a bunch of `Error()` APIs, especially in the context of `componentXXXError`, where they serve a completely different purpose. A comment here indicating:
- the signaler is creating a channel,
- where to find it after
Would help